Importing The Python Libraries


In [ ]:
#Importing the Librarires
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.model_selection import cross_val_score
from sklearn.metrics import classification_report

Load The Dataset

In [ ]:
#Load the Dataset
vaccine = pd.read_csv('h1n1_vaccine_prediction.csv')
vaccine.head()

unique_id  h1n1_worry  h1n1_awareness  antiviral_medication  \
0          0         1.0             0.0                   0.0   
1          1         3.0             2.0                   0.0   
2          2         1.0             1.0                   0.0   
3          3         1.0             1.0                   0.0   
4          4         2.0             1.0                   0.0   

   contact_avoidance  bought_face_mask  wash_hands_frequently  \
0                0.0               0.0                    0.0   
1                1.0               0.0                    1.0   
2                1.0               0.0                    0.0   
3                1.0               0.0                    1.0   
4                1.0               0.0                    1.0   

   avoid_large_gatherings  reduced_outside_home_cont  avoid_touch_face  ...  \
0                     0.0                        1.0               1.0  ...   
1                     0.0                        1.0               1.0  ...   
2                     0.0                        0.0               0.0  ...   
3                     1.0                        0.0               0.0  ...   
4                     1.0                        0.0               1.0  ...   

    race     sex               income_level  marital_status  housing_status  \
0  White  Female              Below Poverty     Not Married             Own   
1  White    Male              Below Poverty     Not Married            Rent   
2  White    Male  <= $75,000, Above Poverty     Not Married             Own   
3  White  Female              Below Poverty     Not Married            Rent   
4  White  Female  <= $75,000, Above Poverty         Married             Own   

           employment                census_msa  no_of_adults  no_of_children  \
0  Not in Labor Force                   Non-MSA           0.0             0.0   
1            Employed  MSA, Not Principle  City           0.0             0.0   
2            Employed  MSA, Not Principle  City           2.0             0.0   
3  Not in Labor Force       MSA, Principle City           0.0             0.0   
4            Employed  MSA, Not Principle  City           1.0             0.0   

   h1n1_vaccine  
0             0  
1             0  
2             0  
3             0  
4             0  

[5 rows x 34 columns]

In [ ]:
vaccine.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26707 entries, 0 to 26706
Data columns (total 34 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   unique_id                  26707 non-null  int64  
 1   h1n1_worry                 26615 non-null  float64
 2   h1n1_awareness             26591 non-null  float64
 3   antiviral_medication       26636 non-null  float64
 4   contact_avoidance          26499 non-null  float64
 5   bought_face_mask           26688 non-null  float64
 6   wash_hands_frequently      26665 non-null  float64
 7   avoid_large_gatherings     26620 non-null  float64
 8   reduced_outside_home_cont  26625 non-null  float64
 9   avoid_touch_face           26579 non-null  float64
 10  dr_recc_h1n1_vacc          24547 non-null  float64
 11  dr_recc_seasonal_vacc      24547 non-null  float64
 12  chronic_medic_condition    25736 non-null  float64
 13  cont_child_undr_6_mnths    25887 non-null  flo

In [ ]:
vaccine.shape

(26707, 34)

In [ ]:
vaccine.describe()

unique_id    h1n1_worry  h1n1_awareness  antiviral_medication  \
count  26707.000000  26615.000000    26591.000000          26636.000000   
mean   13353.000000      1.618486        1.262532              0.048844   
std     7709.791156      0.910311        0.618149              0.215545   
min        0.000000      0.000000        0.000000              0.000000   
25%     6676.500000      1.000000        1.000000              0.000000   
50%    13353.000000      2.000000        1.000000              0.000000   
75%    20029.500000      2.000000        2.000000              0.000000   
max    26706.000000      3.000000        2.000000              1.000000   

       contact_avoidance  bought_face_mask  wash_hands_frequently  \
count       26499.000000      26688.000000           26665.000000   
mean            0.725612          0.068982               0.825614   
std             0.446214          0.253429               0.379448   
min             0.000000          0.000000               0.000000   
25%             0.000000          0.000000               1.000000   
50%             1.000000          0.000000               1.000000   
75%             1.000000          0.000000               1.000000   
max             1.000000          1.000000               1.000000   

       avoid_large_gatherings  reduced_outside_home_cont  avoid_touch_face  \
count             26620.00000               26625.000000      26579.000000   
mean                  0.35864                   0.337315          0.677264   
std                   0.47961                   0.472802          0.467531   
min                   0.00000                   0.000000          0.000000   
25%                   0.00000                   0.000000          0.000000   
50%                   0.00000                   0.000000          1.000000   
75%                   1.00000                   1.000000          1.000000   
max                   1.00000                   1.000000          1.000000   

       ...  has_health_insur  is_h1n1_vacc_effective  is_h1n1_risky  \
count  ...       14433.00000            26316.000000   26319.000000   
mean   ...           0.87972                3.850623       2.342566   
std    ...           0.32530                1.007436       1.285539   
min    ...           0.00000                1.000000       1.000000   
25%    ...           1.00000                3.000000       1.000000   
50%    ...           1.00000                4.000000       2.000000   
75%    ...           1.00000                5.000000       4.000000   
max    ...           1.00000                5.000000       5.000000   

       sick_from_h1n1_vacc  is_seas_vacc_effective  is_seas_risky  \
count         26312.000000            26245.000000   26193.000000   
mean              2.357670                4.025986       2.719162   
std               1.362766                1.086565       1.385055   
min               1.000000                1.000000       1.000000   
25%               1.000000                4.000000       2.000000   
50%               2.000000                4.000000       2.000000   
75%               4.000000                5.000000       4.000000   
max               5.000000                5.000000       5.000000   

       sick_from_seas_vacc  no_of_adults  no_of_children  h1n1_vaccine  
count         26170.000000  26458.000000    26458.000000  26707.000000  
mean              2.118112      0.886499        0.534583      0.212454  
std               1.332950      0.753422        0.928173      0.409052  
min               1.000000      0.000000        0.000000      0.000000  
25%               1.000000      0.000000        0.000000      0.000000  
50%               2.000000      1.000000        0.000000      0.000000  
75%               4.000000      1.000000        1.000000      0.000000  
max               5.000000      3.000000        3.000000      1.000000  

[8 rows x 25 columns]

In [ ]:
vaccine.isnull().sum()   #Checking the Null Values

unique_id                        0
h1n1_worry                      92
h1n1_awareness                 116
antiviral_medication            71
contact_avoidance              208
bought_face_mask                19
wash_hands_frequently           42
avoid_large_gatherings          87
reduced_outside_home_cont       82
avoid_touch_face               128
dr_recc_h1n1_vacc             2160
dr_recc_seasonal_vacc         2160
chronic_medic_condition        971
cont_child_undr_6_mnths        820
is_health_worker               804
has_health_insur             12274
is_h1n1_vacc_effective         391
is_h1n1_risky                  388
sick_from_h1n1_vacc            395
is_seas_vacc_effective         462
is_seas_risky                  514
sick_from_seas_vacc            537
age_bracket                      0
qualification                 1407
race                             0
sex                              0
income_level                  4423
marital_status                1408
housing_status      

EDA To fill The Null Values

In [ ]:
display(vaccine['h1n1_worry'].mode())

0    2.0
dtype: float64

In [ ]:
display (vaccine['h1n1_worry'].median())

2.0

In [ ]:
vaccine['h1n1_worry'].fillna(vaccine['h1n1_worry'].mode()[0], inplace = True)

In [ ]:
vaccine['h1n1_awareness'].unique()

array([ 0.,  2.,  1., nan])

In [ ]:
vaccine['h1n1_awareness'].mode()

0    1.0
dtype: float64

In [ ]:
vaccine['h1n1_awareness'].median()

1.0

In [ ]:
vaccine['h1n1_awareness'].value_counts()

1.0    14598
2.0     9487
0.0     2506
Name: h1n1_awareness, dtype: int64

In [ ]:
vaccine['h1n1_awareness'].fillna(vaccine['h1n1_awareness'].mode()[0], inplace = True)

In [ ]:
display(vaccine['antiviral_medication'].mode())

0    0.0
dtype: float64

In [ ]:
vaccine['antiviral_medication'].fillna(vaccine['antiviral_medication'].mode()[0], inplace = True)

In [ ]:
vaccine['contact_avoidance'].value_counts()

1.0    19228
0.0     7271
Name: contact_avoidance, dtype: int64

In [ ]:
vaccine['contact_avoidance'].fillna(vaccine['contact_avoidance'].mode()[0], inplace = True)

In [ ]:
vaccine['bought_face_mask'].value_counts()

0.0    24847
1.0     1841
Name: bought_face_mask, dtype: int64

In [ ]:
vaccine['bought_face_mask'].fillna(vaccine['bought_face_mask'].mode()[0], inplace = True)

In [ ]:
vaccine['wash_hands_frequently'].fillna(vaccine['wash_hands_frequently'].mode()[0], inplace = True)

In [ ]:
vaccine['avoid_large_gatherings'].fillna(vaccine['avoid_large_gatherings'].mode()[0], inplace = True)

In [ ]:
vaccine['reduced_outside_home_cont'].fillna(vaccine['reduced_outside_home_cont'].mode()[0], inplace = True)

In [ ]:
vaccine['avoid_touch_face'].fillna(vaccine['avoid_touch_face'].mode()[0], inplace = True)

In [ ]:
vaccine['dr_recc_h1n1_vacc'].fillna(vaccine['dr_recc_h1n1_vacc'].mode()[0], inplace = True)
vaccine['dr_recc_seasonal_vacc'].fillna(vaccine['dr_recc_seasonal_vacc'].mode()[0], inplace = True)

In [ ]:
vaccine['chronic_medic_condition'].fillna(vaccine['chronic_medic_condition'].mode()[0], inplace = True)

In [ ]:
vaccine['cont_child_undr_6_mnths'].fillna(vaccine['cont_child_undr_6_mnths'].mode()[0], inplace = True)

In [ ]:
vaccine['is_health_worker'].fillna(vaccine['is_health_worker'].mode()[0], inplace = True)

In [ ]:
vaccine['has_health_insur'].value_counts()

1.0    12697
0.0     1736
Name: has_health_insur, dtype: int64

As no of missing values is 12274 which is almost equal to no of persons who have health insurance. Therefore, it is not a good option to replace missing values by mode as there is no surity whether they have insurance. Hence, putting them in another category named '2.0' which means 'data unknown'.

In [ ]:
vaccine['has_health_insur'].fillna(2.0, inplace = True)

In [ ]:
vaccine['is_h1n1_vacc_effective'].fillna(vaccine['is_h1n1_vacc_effective'].mode()[0], inplace = True)
vaccine['is_seas_vacc_effective'].fillna(vaccine['is_seas_vacc_effective'].mode()[0], inplace = True)

In [ ]:
vaccine['is_h1n1_risky'].fillna(vaccine['is_h1n1_risky'].mode()[0], inplace = True)
vaccine['is_seas_risky'].fillna(vaccine['is_seas_risky'].mode()[0], inplace = True)

In [ ]:
vaccine['sick_from_h1n1_vacc'].fillna(vaccine['sick_from_h1n1_vacc'].mode()[0], inplace = True)
vaccine['sick_from_seas_vacc'].fillna(vaccine['sick_from_seas_vacc'].mode()[0], inplace = True)

In [ ]:
vaccine['qualification'].fillna(vaccine['qualification'].mode()[0], inplace = True)

In [ ]:
vaccine['income_level'].value_counts()

<= $75,000, Above Poverty    12777
> $75,000                     6810
Below Poverty                 2697
Name: income_level, dtype: int64

No.of missing values is 4423, better to create a new category as 'Unknown' as this will not skew the data and info will not be lost

In [ ]:
vaccine['income_level'].fillna('Unknown', inplace = True)  

In [ ]:
vaccine['marital_status'].fillna(vaccine['marital_status'].mode()[0], inplace = True)

In [ ]:
vaccine['housing_status'].fillna(vaccine['housing_status'].mode()[0], inplace = True)

In [ ]:
vaccine['employment'].fillna(vaccine['employment'].mode()[0], inplace = True)

In [ ]:
vaccine['no_of_adults'].fillna(vaccine['no_of_adults'].mode()[0], inplace = True)
vaccine['no_of_children'].fillna(vaccine['no_of_children'].mode()[0], inplace = True)

In [ ]:
vaccine.isnull().sum()   #Result of EDA all Null Values are Filled or Removed

unique_id                    0
h1n1_worry                   0
h1n1_awareness               0
antiviral_medication         0
contact_avoidance            0
bought_face_mask             0
wash_hands_frequently        0
avoid_large_gatherings       0
reduced_outside_home_cont    0
avoid_touch_face             0
dr_recc_h1n1_vacc            0
dr_recc_seasonal_vacc        0
chronic_medic_condition      0
cont_child_undr_6_mnths      0
is_health_worker             0
has_health_insur             0
is_h1n1_vacc_effective       0
is_h1n1_risky                0
sick_from_h1n1_vacc          0
is_seas_vacc_effective       0
is_seas_risky                0
sick_from_seas_vacc          0
age_bracket                  0
qualification                0
race                         0
sex                          0
income_level                 0
marital_status               0
housing_status               0
employment                   0
census_msa                   0
no_of_adults                 0
no_of_ch

In [ ]:
vaccine.head()

unique_id  h1n1_worry  h1n1_awareness  antiviral_medication  \
0          0         1.0             0.0                   0.0   
1          1         3.0             2.0                   0.0   
2          2         1.0             1.0                   0.0   
3          3         1.0             1.0                   0.0   
4          4         2.0             1.0                   0.0   

   contact_avoidance  bought_face_mask  wash_hands_frequently  \
0                0.0               0.0                    0.0   
1                1.0               0.0                    1.0   
2                1.0               0.0                    0.0   
3                1.0               0.0                    1.0   
4                1.0               0.0                    1.0   

   avoid_large_gatherings  reduced_outside_home_cont  avoid_touch_face  ...  \
0                     0.0                        1.0               1.0  ...   
1                     0.0                        1.0               1.0  ...   
2                     0.0                        0.0               0.0  ...   
3                     1.0                        0.0               0.0  ...   
4                     1.0                        0.0               1.0  ...   

    race     sex               income_level  marital_status  housing_status  \
0  White  Female              Below Poverty     Not Married             Own   
1  White    Male              Below Poverty     Not Married            Rent   
2  White    Male  <= $75,000, Above Poverty     Not Married             Own   
3  White  Female              Below Poverty     Not Married            Rent   
4  White  Female  <= $75,000, Above Poverty         Married             Own   

           employment                census_msa  no_of_adults  no_of_children  \
0  Not in Labor Force                   Non-MSA           0.0             0.0   
1            Employed  MSA, Not Principle  City           0.0             0.0   
2            Employed  MSA, Not Principle  City           2.0             0.0   
3  Not in Labor Force       MSA, Principle City           0.0             0.0   
4            Employed  MSA, Not Principle  City           1.0             0.0   

   h1n1_vaccine  
0             0  
1             0  
2             0  
3             0  
4             0  

[5 rows x 34 columns]

In [ ]:
vaccine["race"].unique()

array(['White', 'Black', 'Other or Multiple', 'Hispanic'], dtype=object)

Creating Dummies


In [ ]:
race = pd.get_dummies(vaccine["race"],drop_first=True)
race.head()

Hispanic  Other or Multiple  White
0         0                  0      1
1         0                  0      1
2         0                  0      1
3         0                  0      1
4         0                  0      1

In [ ]:
sex = pd.get_dummies(vaccine["sex"],drop_first = True)
sex.head()

Male
0     0
1     1
2     1
3     0
4     0

In [ ]:
income = pd.get_dummies(vaccine["income_level"],drop_first= True)
income.head()

> $75,000  Below Poverty  Unknown
0          0              1        0
1          0              1        0
2          0              0        0
3          0              1        0
4          0              0        0

In [ ]:
age = pd.get_dummies(vaccine["age_bracket"],drop_first= True)
age.head()

35 - 44 Years  45 - 54 Years  55 - 64 Years  65+ Years
0              0              0              1          0
1              1              0              0          0
2              0              0              0          0
3              0              0              0          1
4              0              1              0          0

In [ ]:
vaccine.drop("census_msa",axis = 1)

unique_id  h1n1_worry  h1n1_awareness  antiviral_medication  \
0              0         1.0             0.0                   0.0   
1              1         3.0             2.0                   0.0   
2              2         1.0             1.0                   0.0   
3              3         1.0             1.0                   0.0   
4              4         2.0             1.0                   0.0   
...          ...         ...             ...                   ...   
26702      26702         2.0             0.0                   0.0   
26703      26703         1.0             2.0                   0.0   
26704      26704         2.0             2.0                   0.0   
26705      26705         1.0             1.0                   0.0   
26706      26706         0.0             0.0                   0.0   

       contact_avoidance  bought_face_mask  wash_hands_frequently  \
0                    0.0               0.0                    0.0   
1                    1.0               0.0                    1.0   
2                    1.0               0.0                    0.0   
3                    1.0               0.0                    1.0   
4                    1.0               0.0                    1.0   
...                  ...               ...                    ...   
26702                1.0               0.0                    0.0   
26703                1.0               0.0                    1.0   
26704                1.0               1.0                    1.0   
26705                0.0               0.0                    0.0   
26706                1.0               0.0                    0.0   

       avoid_large_gatherings  reduced_outside_home_cont  avoid_touch_face  \
0                         0.0                        1.0               1.0   
1                         0.0                        1.0               1.0   
2                         0.0                        0.0               0.0   
3                         1.0                        0.0               0.0   
4                         1.0                        0.0               1.0   
...                       ...                        ...               ...   
26702                     0.0                        1.0               0.0   
26703                     0.0                        0.0               0.0   
26704                     1.0                        0.0               1.0   
26705                     0.0                        0.0               1.0   
26706                     0.0                        0.0               0.0   

       ...     qualification      race     sex               income_level  \
0      ...        < 12 Years     White  Female              Below Poverty   
1      ...          12 Years     White    Male              Below Poverty   
2      ...  College Graduate     White    Male  <= $75,000, Above Poverty   
3      ...          12 Years     White  Female              Below Poverty   
4      ...      Some College     White  Female  <= $75,000, Above Poverty   
...    ...               ...       ...     ...                        ...   
26702  ...      Some College     White  Female  <= $75,000, Above Poverty   
26703  ...  College Graduate     White    Male  <= $75,000, Above Poverty   
26704  ...      Some College     White  Female                    Unknown   
26705  ...      Some College  Hispanic  Female  <= $75,000, Above Poverty   
26706  ...      Some College     White    Male  <= $75,000, Above Poverty   

       marital_status  housing_status          employment  no_of_adults  \
0         Not Married             Own  Not in Labor Force           0.0   
1         Not Married            Rent            Employed           0.0   
2         Not Married             Own            Employed           2.0   
3         Not Married            Rent  Not in Labor Force           0.0   
4             Married             Own            Employed           1.0   
...               ...    

In [ ]:
married = pd.get_dummies(vaccine["marital_status"],drop_first = True)
married.head()

Not Married
0            1
1            1
2            1
3            1
4            0

In [ ]:
house = pd.get_dummies(vaccine["housing_status"],drop_first = True)
house.head()

Rent
0     0
1     1
2     0
3     1
4     0

In [ ]:
employee= pd.get_dummies(vaccine["employment"],drop_first = True)
employee.head()

Not in Labor Force  Unemployed
0                   1           0
1                   0           0
2                   0           0
3                   1           0
4                   0           0

In [ ]:
vaccine = pd.concat([vaccine,race,sex,income,age,married,house,employee],axis=1)

In [ ]:
vaccine.columns

Index(['unique_id', 'h1n1_worry', 'h1n1_awareness', 'antiviral_medication',
       'contact_avoidance', 'bought_face_mask', 'wash_hands_frequently',
       'avoid_large_gatherings', 'reduced_outside_home_cont',
       'avoid_touch_face', 'dr_recc_h1n1_vacc', 'dr_recc_seasonal_vacc',
       'chronic_medic_condition', 'cont_child_undr_6_mnths',
       'is_health_worker', 'has_health_insur', 'is_h1n1_vacc_effective',
       'is_h1n1_risky', 'sick_from_h1n1_vacc', 'is_seas_vacc_effective',
       'is_seas_risky', 'sick_from_seas_vacc', 'age_bracket', 'qualification',
       'race', 'sex', 'income_level', 'marital_status', 'housing_status',
       'employment', 'census_msa', 'no_of_adults', 'no_of_children',
       'h1n1_vaccine', 'Hispanic', 'Other or Multiple', 'White', 'Male',
       '> $75,000', 'Below Poverty', 'Unknown', '35 - 44 Years',
       '45 - 54 Years', '55 - 64 Years', '65+ Years', 'Not Married', 'Rent',
       'Not in Labor Force', 'Unemployed'],
      dtype='object')

In [ ]:
x=vaccine.drop(columns = ["h1n1_vaccine","census_msa","income_level","race","sex","qualification","income_level","age_bracket","marital_status","housing_status","employment"])
y=vaccine["h1n1_vaccine"]

In [ ]:
vaccine.columns

Index(['unique_id', 'h1n1_worry', 'h1n1_awareness', 'antiviral_medication',
       'contact_avoidance', 'bought_face_mask', 'wash_hands_frequently',
       'avoid_large_gatherings', 'reduced_outside_home_cont',
       'avoid_touch_face', 'dr_recc_h1n1_vacc', 'dr_recc_seasonal_vacc',
       'chronic_medic_condition', 'cont_child_undr_6_mnths',
       'is_health_worker', 'has_health_insur', 'is_h1n1_vacc_effective',
       'is_h1n1_risky', 'sick_from_h1n1_vacc', 'is_seas_vacc_effective',
       'is_seas_risky', 'sick_from_seas_vacc', 'age_bracket', 'qualification',
       'race', 'sex', 'income_level', 'marital_status', 'housing_status',
       'employment', 'census_msa', 'no_of_adults', 'no_of_children',
       'h1n1_vaccine', 'Hispanic', 'Other or Multiple', 'White', 'Male',
       '> $75,000', 'Below Poverty', 'Unknown', '35 - 44 Years',
       '45 - 54 Years', '55 - 64 Years', '65+ Years', 'Not Married', 'Rent',
       'Not in Labor Force', 'Unemployed'],
      dtype='object')

In [ ]:
vaccine["age_bracket"].unique()

array(['55 - 64 Years', '35 - 44 Years', '18 - 34 Years', '65+ Years',
       '45 - 54 Years'], dtype=object)

In [ ]:
x_train,x_test, y_train,y_test = train_test_split(x,y,test_size= 0.3,random_state = 42)

In [ ]:
lg= LogisticRegression()

In [ ]:
lg.fit(x_train,y_train)

LogisticRegression()

In [ ]:
prediction = lg.predict(x_test)

In [ ]:
from sklearn.metrics import classification_report

In [ ]:
classification_report(y_test,prediction)

'              precision    recall  f1-score   support\n\n           0       0.85      0.94      0.89      6319\n           1       0.63      0.39      0.49      1694\n\n    accuracy                           0.82      8013\n   macro avg       0.74      0.67      0.69      8013\nweighted avg       0.81      0.82      0.81      8013\n'

In [ ]:
from sklearn.metrics import confusion_matrix

In [ ]:
confusion_matrix(y_test,prediction)

array([[5935,  384],
       [1028,  666]])